src/sage/schemes/curves/plane_curve_arrangement.py

```diff
@@ -0,0 +1,1246 @@
+            arg = (args[0], )
+        elif len(args) == 1:
+            arg = tuple(args[0])
+        else:
+            arg = tuple(args)
+        # process keyword arguments
+        if len(kwds) > 0:
+            raise ValueError('unknown keyword argument')
+        # process positional arguments
+        ambient_space = self.ambient_space()
+        R = ambient_space.coordinate_ring()
+        curves = ()
+        for h in arg:
+            try:
+                ambient = h.ambient_space()
+                if ambient == ambient_space:
+                    curves += (h, )
+                else:
+                    return None
+            except AttributeError:
+                try:
+                    h = R(h)
+                    curves += (Curve(h), )
+                except TypeError:
+                    return None
+        return self.element_class(self, curves)
+
+    def _an_element_(self):
+        """
+        Return an element of ``self``.
+
+        TESTS::
+
+            sage: H.<t, s> = AffinePlaneCurveArrangements(QQ)
+            sage: H._an_element_()
+            Arrangement (t) in Affine Space of dimension 2 over Rational Field
+            sage: H.<t, s, r> = ProjectivePlaneCurveArrangements(QQ)
+            sage: H._an_element_()
+            Arrangement (t) in Projective Space of dimension 2 over Rational Field
+        """
+        x = self.gen(0)
+        return self(x)
+
+    @cached_method
+    def ngens(self):
+        """
+        Return the number of variables, i.e. 2 or 3, kept for completness.
+
+        OUTPUT:
+
+        An integer, 2 or 3, depending if the arrangement is projective or affine.
+
+        EXAMPLES::
+
+            sage: L.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: L.ngens()
+            2
+            sage: L.<x, y, z> = ProjectivePlaneCurveArrangements(QQ)
+            sage: L.ngens()
+            3
+        """
+        return len(self._names)
+
+    @cached_method
+    def gens(self):
+        """
+        Return the coordinates.
+
+        OUTPUT:
+
+        A tuple of linear expressions, one for each linear variable.
+
+        EXAMPLES::
+
+            sage: L = AffinePlaneCurveArrangements(QQ, ('x', 'y'))
+            sage: L.gens()
+            (x, y)
+            sage: L = ProjectivePlaneCurveArrangements(QQ, ('x', 'y', 'z'))
+            sage: L.gens()
+            (x, y, z)
+        """
+        return self.ambient_space().gens()
+
+    def gen(self, i):
+        """
+        Return the `i`-th coordinate.
+
+        INPUT:
+
+        - ``i`` -- integer
+
+        OUTPUT:
+
+        A variable.
+
+        EXAMPLES::
+
+            sage: L.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: L.gen(1)
+            y
+            sage: L.<x, y, z> = ProjectivePlaneCurveArrangements(QQ)
+            sage: L.gen(2)
+            z
+        """
+        return self.gens()[i]
+
+
+class ProjectivePlaneCurveArrangements(AffinePlaneCurveArrangements):
```

Comment:
In principle one may think that an affine arrangement is also projective including the line at infinity, but there are affine isomorphisms which do not preserve it. It is more or less like the two classes `AffineCurve` and `ProjectiveCurve`.

Reply:
But you need to add something to it to get the equivalence, right? If so, this is a slight abuse (which can be factored out into a common ABC = (Abstract Base Class)).

This is a bit surprising to me as I am used to building projective objects from affine objects.

Reply:
I do not know what to say. I am open to suggestions; in `projective_curve-py` the class `ProjectiveCurve` is also created.

Reply:
I don't see your point about affine vs projective curves. They basically have completely different MROs until you get to the generic classes (e.g., `Curve_generic`).

Here, you are declaring that an affine plane curve arrangement *is* an a projective plane curve arrangement. Not equivalent by changing/adding something canonically, exactly is because `isinstance(PCA, AffinePlaneCurveArrangement)` will be `True` when `PCA` is an `ProjectivePlaneCurveArragement`.

Of course, as you said, you can embed affine space into projective space by adding a point. However, this is the opposite direction of your subclassing. I don't quite get how you get a line from a point either.

Reply:
Both classes share a lot of properties, this is why `PCA` inherits from `ACA` (`AffinePlaneCurveArrangement`). But `fundamental_group` and `meridians` have a different behavior (as `ambient_space`, etc). I though that it was not worth to duplicate code.

Reply:
I completely agree, and so I was suggesting you should have a common ABC for both classes. Although re-reading what I wrote, this was not very clear; sorry for that.